ember-core/shared/src/test/scala/org/http4s/ember/core/h2/H2StreamSuite.scala

```diff
@@ -0,0 +1,110 @@
+package org.http4s.ember.core.h2
+
+import cats.syntax.all._
+import cats.effect.std.Queue
+import org.http4s.{
+  EntityEncoder,
+  Headers,
+  Http4sSuite,
+  HttpVersion,
+  Message,
+  Request,
+  Response,
+  Status,
+}
+import org.http4s.ember.core.h2.H2Frame.Settings.SettingsMaxFrameSize
+import cats.effect.{Deferred, IO, Ref}
+import fs2.Chunk
+import org.http4s.ember.core.h2.H2Frame
+import org.scalacheck.Gen.choose
+import org.scalacheck.effect.PropF.forAllF
+import org.typelevel.log4cats.Logger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+import scodec.bits.ByteVector
+
+class H2StreamSuite extends Http4sSuite {
+  test("H2Stream should split data frames as defined by client settings") {
+    val defaultSettings = H2Frame.Settings.ConnectionSettings.default
+
+    forAllF(choose(16 << 10, 64 << 10)) { frameSize =>
+      val config = defaultSettings.copy(
+        maxFrameSize = SettingsMaxFrameSize(frameSize)
+      )
+
+      def testMessageSize(
+          outgoing: Queue[IO, Chunk[H2Frame]],
+          stream: H2Stream[IO],
+          messageSize: Int,
+          frames: Int,
+      ) = {
+        val sample = Response[IO](Status.Ok, HttpVersion.`HTTP/2`).withEntity("0" * messageSize)
+
+        for {
+          _ <- stream.sendMessageBody(sample)
+          chunks <- outgoing
+            .tryTakeN(None)
+            .map(_.flatMap(_.toList).collect { case H2Frame.Data(_, data, _, _) =>
+              data
+            })
+          _ <- assertIO(IO(chunks.size), frames)
+          _ <- assertIO(IO(chunks.map(_.size).sum), messageSize.toLong)
+          _ <- assertIO_(
+            chunks.map(c => assertIOBoolean(IO(c.size < config.maxFrameSize.frameSize))).sequence_
+          )
```

Comment:
Should this be `<=`? Also I'm not sure why it needs to be wrapped in `assertIO_(...)`. 
```suggestion
          _ <- chunks.traverse_(c => IO(assert(clue(c.size) <= clue(config.maxFrameSize.frameSize)))))
```